python/cudf/cudf/utils/ioutils.py

```diff
@@ -1391,3 +1508,182 @@ def _ensure_filesystem(passed_filesystem, path):
+        )
+
+    # Calculate total file size
+    if file_like:
+        file_size = path_or_fob.size
+    file_size = file_size or fs.size(path_or_fob)
+
+    # Check if a direct read makes the most sense
+    if not byte_ranges and bytes_per_thread >= file_size:
+        if file_like:
+            return path_or_fob.read()
+        else:
+            return fs.open(path_or_fob, mode=mode, cache_type="bytes").read()
+
+    # Threaded read into "dummy" buffer
+    buf = np.zeros(file_size, dtype="b")
+    if byte_ranges:
+
+        # Optimize/merge the ranges
+        byte_ranges = _merge_ranges(
+            byte_ranges, max_block=bytes_per_thread, max_gap=max_gap,
+        )
+
+        # Call multi-threaded data transfer of
+        # remote byte-ranges to local buffer
+        _read_byte_ranges(
+            path_or_fob, byte_ranges, buf, fs=fs, **kwargs,
+        )
+
+        # Add Header & Footer bytes
+        if footer is not None:
+            footer_size = len(footer)
+            buf[-footer_size:] = np.frombuffer(
+                footer[-footer_size:], dtype="b"
+            )
+
+        # Add parquet magic bytes (optional)
+        if add_par1_magic:
+            buf[:4] = np.frombuffer(b"PAR1", dtype="b")
+            if footer is None:
+                buf[-4:] = np.frombuffer(b"PAR1", dtype="b")
+
+    else:
+        byte_ranges = [
+            (b, min(bytes_per_thread, file_size - b))
+            for b in range(0, file_size, bytes_per_thread)
+        ]
+        _read_byte_ranges(
+            path_or_fob, byte_ranges, buf, fs=fs, **kwargs,
+        )
+
+    return buf.tobytes()
+
+
+def _merge_ranges(byte_ranges, max_block=256_000_000, max_gap=64_000):
+    # Simple utility to merge small/adjacent byte ranges
+    new_ranges = []
+    if not byte_ranges:
+        # Early return
+        return new_ranges
+
+    offset, size = byte_ranges[0]
+    for (new_offset, new_size) in byte_ranges[1:]:
+        gap = new_offset - (offset + size)
+        if gap > max_gap or (size + new_size + gap) > max_block:
+            # Gap is too large or total read is too large
+            new_ranges.append((offset, size))
+            offset = new_offset
+            size = new_size
+            continue
+        size += new_size + gap
+    new_ranges.append((offset, size))
+    return new_ranges
+
+
+def _assign_block(fs, path_or_fob, local_buffer, offset, nbytes):
+    if fs is None:
+        # We have an open fsspec file object
+        path_or_fob.seek(offset)
+        local_buffer[offset : offset + nbytes] = np.frombuffer(
+            path_or_fob.read(nbytes), dtype="b",
+        )
+    else:
+        # We have an fsspec filesystem and a path
+        with fs.open(path_or_fob, mode="rb", cache_type="none") as fob:
+            fob.seek(offset)
+            local_buffer[offset : offset + nbytes] = np.frombuffer(
+                fob.read(nbytes), dtype="b",
+            )
+
```

Comment:
@leiterenato - Perhaps you can comment on the most optimal API to read a specific set of bytes in gcs?

Hopefully, we can add those optimizations directly to [gcsfs](https://gcsfs.readthedocs.io/en/latest/) so that a simple `fs.read_block(...)` call would be optimal here. Note that I am using seek/read for now, since `read_block` will actually open the file with read-ahead caching, and *then* call seek/read.

Reply:
@rjzamora 
Currently the tool `gcloud alpha storage cp`  has the most optimized implementation.
[Download link.](https://dl.google.com/dl/cloudsdk/channels/rapid/downloads/google-cloud-sdk-356.0.0-linux-x86_64.tar.gz)
The source code is in this directory: lib/googlecloudsdk/command_lib/storage/tasks/cp/.
[Blog post with more information.](https://cloud.google.com/blog/products/storage-data-transfer/new-gcloud-storage-enables-super-fast-data-transfers)

Reply:
gcsfs supports reading the single block using `cat`, possibly for multiple blocks in multiple files concurrently.  We just need to push https://github.com/intake/filesystem_spec/pull/744 over the line for is to be available in all async implementations.